Run tests on request data

mjethani opened the pull request:
This runs the tests on the request data set from https://github.com/mjethani/scaling-palm-tree

mjethani commented on platform/npm/test.js:
```diff
@@ -32,7 +32,7 @@ import { promisify } from 'util';
 
 async function spawnMocha() {
-    await promisify(spawn)('mocha', [ '--experimental-vm-modules', '--no-warnings', 'tests' ], { stdio: [ 'inherit', 'inherit', 'inherit' ] });
+    await promisify(spawn)('mocha', [ '--experimental-vm-modules', '--no-warnings', 'tests', '--reporter', 'progress' ], { stdio: [ 'inherit', 'inherit', 'inherit' ] });
```
@gorhill we need a different reporter because otherwise there's too much output. This one will only show something if a test fails.

mjethani commented on platform/npm/tests/request-data.js:
```diff
@@ -0,0 +1,79 @@
+const require = createRequire(import.meta.url);
+
+const requests = require('scaling-palm-tree/requests.json');
+const results = require('../results.json');
```
The `results.json` was generated basically by running this same code but saving the results to a file.

mjethani commented on platform/npm/tests/request-data.js:
```diff
@@ -0,0 +1,79 @@
+const requests = require('scaling-palm-tree/requests.json');
+const results = require('../results.json');
+
+const typeMap = {
```
Mapping types here.

mjethani commented on platform/npm/tests/request-data.js:
```diff
@@ -0,0 +1,79 @@
+
+        const expected = results[i];
+
+        it(`should ${expected === 1 ? 'block' : 'allow'} ${request.type} URL ${request.url} from origin ${request.originURL}`, () => {
```
If a test fails it should say something like "should block script URL https://example.com/foo.js from origin https://example.net/index.html" followed by the stack trace.

mjethani commented:
@gorhill try to run this and see how it works.

gorhill commented:
I think we will need to include some of uBO's own filter lists otherwise this would leave out a lot of untested code paths in the filtering engine because of filter options not used in EasyList, such as `important`, `denyallow`, and so on.

mjethani commented on tools/make-npm.sh:
```diff
@@ -35,10 +35,11 @@ node -pe "JSON.stringify(fs.readFileSync('$THIRDPARTY/easyprivacy.txt', 'utf8'))
-cp    platform/npm/.*.json $DES/
-cp    platform/npm/*.js    $DES/
-cp -R platform/npm/tests   $DES/
+cp    platform/npm/.npmignore  $DES/
```
Sorry, I have force-pushed a couple of times. There is a `.npmignore` file now and it ignores `results.json` and `tests/`.

mjethani commented:
> I think we will need to include some of uBO's own filter lists otherwise this would leave out a lot of untested code paths in the filtering engine because of filter options not used in EasyList, such as `important`, `denyallow`, and so on.

I'll do a follow-up patch to include more lists, and do let me know which ones to include.

gorhill commented:
> I'll do a follow-up patch to include more lists, and do let me know which ones to include.

I would say we should match the one used by default in uBO, so just like <https://github.com/gorhill/uBlock/blob/514143cd805e31235b6b0c9f21183707df4d9b7f/platform/dig/snfe.js#L131-L154> -- hence why in the end I ended up using <https://github.com/gorhill/uBlock/blob/514143cd805e31235b6b0c9f21183707df4d9b7f/tools/make-assets.sh> to add the lists.

mjethani commented:
> I would say we should match the one used by default in uBO

Does this mean you'd like to ship these lists with the Node.js package too?

Else we put them in a separate directory and add that to `.npmignore`.

gorhill commented:
> Does this mean you'd like to ship these lists with the Node.js package too?

Not sure what to do really, I am tempted to say we ship no lists at all (they would be all outdated anyways), and use them just for the test suites.

mjethani commented:
> … I am tempted to say we ship no lists at all (they would be all outdated anyways), and use them just for the test suites.

Sounds good to me. In that case, we could just load the lists from their original locations. Anyone doing development still needs to clone the entire repo anyway.

mjethani commented on tools/make-npm.sh:
```diff
@@ -18,6 +18,7 @@ fi
 rm -rf $DES
 
 ./tools/make-nodejs.sh $DES
+./tools/make-assets.sh $DES
```
I'm just running `./tools/make-assets.sh` for consistency. The `assets` directory is in `.npmignore`.

gorhill replied:
A problem I see left is that the results depends on the filter lists, and these change constantly, so this will inevitable cause spurious errors -- I gather `results.json.gz` is not meant to be constantly updated? In such case maybe we should allow a second argument to `make-assets.sh` which is the commit hash to use when importing from uAssets.

mjethani replied:
Yes, this is a problem, and this is why even in the Cliqz benchmarks we don't want to update the lists too often (maybe once every few weeks at most).

Now the problem with the commit hash is that if you have a shallow clone you don't have the history.

What about if we just copied the lists over for testing purposes every once in a while?

mjethani replied:
> What about if we just copied the lists over for testing purposes every once in a while?

By this I mean that we keep separate copies for the tests right in the `platform/npm` directory, perhaps under something like a `test-data` subdirectory.

gorhill replied:
Copied somewhere in `platform/npm`?

mjethani replied:
>Copied somewhere in `platform/npm`?

Essentially instead of a `results.json.gz` we have a `test-data.tar.gz` and it includes everything, which we then unbundle as part of the build script.

gorhill replied:
Bundle sounds good :+1:

mjethani replied:
Done.

mjethani commented on platform/npm/tests/request-data.js:
```diff
@@ -0,0 +1,107 @@
+
+        engine = await StaticNetFilteringEngine.create();
+
+        await engine.useLists([
```
@gorhill based on this comment: https://github.com/gorhill/uBlock/pull/3828#issuecomment-899120180

mjethani commented:
@gorhill `results.json` could be a binary file. If a bit is 0, the result is 0; if it's 1, the next bit indicates if the result is 1 or 2. Most results will be 0.

gorhill commented:
> The point of this would be to save space in the repo.

Or to keep it simpler and avoid further external dependencies, maybe just uze nodejs's [zlib.gunzipSync](https://nodejs.org/dist/latest-v14.x/docs/api/zlib.html#zlib_zlib_gunzipsync_buffer_options)?

mjethani commented:
> > The point of this would be to save space in the repo.
> 
> Or to keep it simpler and avoid further external dependencies, maybe just uze nodejs's [zlib.gunzipSync](https://nodejs.org/dist/latest-v14.x/docs/api/zlib.html#zlib_zlib_gunzipsync_buffer_options)?

This seems like the best way to do it.

Force-pushed.

mjethani commented:
@gorhill the contents of `results.json.gz` were wrong because of what I said in https://github.com/gorhill/uBlock/commit/087da6407d68e95eafa7aa846a5a984e6d20980f#r54947770

Now I've fixed it and also simplified `make-npm.sh`.

Force-pushed.

mjethani commented:
```console
$ git clone --depth=1 https://github.com/gorhill/uBlock.git
Cloning into 'uBlock'...
remote: Enumerating objects: 670, done.
remote: Counting objects: 100% (670/670), done.
remote: Compressing objects: 100% (535/535), done.
remote: Total 670 (delta 121), reused 355 (delta 52), pack-reused 0
Receiving objects: 100% (670/670), 13.81 MiB | 5.48 MiB/s, done.
Resolving deltas: 100% (121/121), done.
$ du -sh uBlock
 37M	uBlock
$ 
```

This is _before_ this patch.

This patch adds ~1M.

I think it's worth it considering that errors such as the one I mentioned earlier (https://github.com/gorhill/uBlock/commit/087da6407d68e95eafa7aa846a5a984e6d20980f#r54947770) would be immediately caught by these tests.

gorhill commented:
What would be the best way to also have a WASM run for the tests? Mainly it's just a matter of calling `enableWASM()` before all tests but I am not sure where this would go.

mjethani commented:
> What would be the best way to also have a WASM run for the tests? Mainly it's just a matter of calling `enableWASM()` before all tests but I am not sure where this would go.

I was looking into it and unfortunately if the engine is loaded via `createWorld()` it doesn't match correctly. I'll have to dig deeper, but mostly it has something to do with a missing global or incorrect loading of JSON files.

This is what I had in mind:

```js
describe('Request data', () => {
    for ( let wasm of [ false, true ] ) {
        context(`${wasm ? 'Wasm on' : 'Wasm off'}`, () => {
            // call enableWASM() if wasm is true
        });
    }
});
```

The above will run the tests twice, once without Wasm and once with Wasm, each in a separate world. I'm not sure that the separate world is needed for now, but you tell me.

In any case, the `createWorld()` implementation needs to be fixed and I'll look into it.

For now we could also just run all the tests with Wasm enabled.

mjethani commented:
> I'm not sure that the separate world is needed for now, but you tell me.

Can we run all the tests, then call `enableWASM()`, and then run all the tests one more time? Not really, right? Because the PSL is already initialized and that's without Wasm.

mjethani commented:
> For now we could also just run all the tests with Wasm enabled.

```diff
diff --git a/platform/npm/tests/request-data.js b/platform/npm/tests/request-data.js
index 963b0aec5..0a11ea05f 100644
--- a/platform/npm/tests/request-data.js
+++ b/platform/npm/tests/request-data.js
@@ -62,7 +62,9 @@ describe('Request data', () => {
     let engine = null;
 
     before(async () => {
-        const { StaticNetFilteringEngine } = await import('../index.js');
+        const { StaticNetFilteringEngine, enableWASM } = await import('../index.js');
+
+        await enableWASM();
 
         engine = await StaticNetFilteringEngine.create();
 
```

gorhill commented:
`enableWASM()` can be called at any time -- the typed arrays used by js are transferred into wasm memory when wasm is enabled, so it should work if being called for a second run, there is no way to go back to no-wasm though once it's enabled.

In any case, I can always test wasm paths with dig platform for now, so it's not a priority.

gorhill commented:
> For now we could also just run all the tests with Wasm enabled.

Best to stick to JS if we are going to test one mode only -- the JS code paths is still the most used given the user base of Chromium users vs. Firefox users.
